src/orchestrator-kubernetes/src/lib.rs

```diff
@@ -1270,31 +1272,51 @@ impl OrchestratorWorker {
 
     async fn run(mut self) {
         while let Some(cmd) = self.command_rx.recv().await {
-            use WorkerCommand::*;
-            let result = match cmd {
-                EnsureService { desc } => self.ensure_service(desc).await,
-                DropService { name } => self.drop_service(&name).await,
-                ListServices {
-                    namespace,
-                    result_tx,
-                } => {
-                    let result = self.list_services(&namespace).await;
-                    let _ = result_tx.send(result);
-                    Ok(())
-                }
-                FetchServiceMetrics {
-                    name,
-                    info,
-                    result_tx,
-                } => {
-                    let result = self.fetch_service_metrics(&name, &info).await;
-                    let _ = result_tx.send(result);
-                    Ok(())
+            self.handle_command(cmd).await;
+        }
+    }
+
+    /// Handle a worker command.
+    ///
+    /// If handling the command fails, it is automatically retried. All command handlers return
+    /// [`K8sError`], so we can reasonably assume that a failure is caused by issues communicating
+    /// with the K8S server and that retrying resolves them eventually.
+    async fn handle_command(&mut self, cmd: WorkerCommand) {
+        macro_rules! with_retry {
+            ($fut:expr, $cmd_type:expr) => {
+                loop {
+                    match $fut.await {
+                        Ok(result) => break result,
+                        Err(error) => {
+                            tracing::error!(
+                                cmd_type = $cmd_type,
+                                "orchestrator API call failed: {error}"
+                            );
+                            tokio::time::sleep(Duration::from_secs(1)).await;
+                        }
+                    }
                 }
             };
+        }
```

Comment:
I tried using the `Retry` helper from `mz-ore` here, but got stuck on async borrowing issues. Also I couldn't implement the retry logic as a loop around `handle_command` because `ListServices` contains a `oneshot` sender and needs to own that to be able to send the result, so the retry needs to be limited to the `list_services` call.

Reply:
Not sure exactly what borrowing issues you were running into, but there are at least two tricks to working around them.

The first trick is to pass state via [`retry_async_with_state`](https://dev.materialize.com/api/rust/mz_ore/retry/struct.Retry.html#method.retry_async_with_state). The idea is that you specify state to be passed in to every invocation of the retried function. That can help work around borrowing issues if each invocation needs exclusive access to an object. The downside is that you need to return the state on every invocation, which means you can't ergonomically use `?` any more.

The second trick is to ditch the closure and use [`into_retry_stream`]:

```rust
let retry = mz_ore::retry::Retry::new().into_retry_stream();
let mut result = None;
while retry.next().await.is_some() {
    result = do_operation();
}
result.expect("always called once")
```

That's the more nuclear option, as you don't get any nice automatic handling of errors, but you do get the exponential backoff at least.

Reply:
> Not sure exactly what borrowing issues you were running into

Well, I also don't remember and it seems to just work on the second try 😅 But thanks for explaining, I'm sure this will be useful down the road!